Merge service accounts for indexer and service (#2143) (#2210)
DataBiosphere/azul: UPGRADING.rst

reverted. This is all fairly informal and loosely defined. Hopefully we won't
have too many entries in this file.

#2143 Merge service accounts for indexer and service
============================================================

The following directions need to be followed for each deployment that needs to

> merge the commit.

Reviewer: Deployments don't merge commits.

Author: `s/deployment/branch`

1.  Let someone who can administer authorization of SAM/TDR know of the desired

>     change in service account via Slack. Note performing these steps with out

Reviewer: Be more specific. This isn't some ominous "change in the service account". The SA is being renamed. Explain what exactly does this person needs to do.

    the manual authorization can lead to deployment errors during
    ``make deploy``.

2.  Before upgrading to this commit, remove the old

>     shared_credentials value in the AWS Secrets Manager, run ::

Reviewer: TMI. The reader does not need to know about shared_credentials?

Author: This section was not needed.

        make -C terraform validate

>         cd terraform && terraform destroy -target=google_service_account.indexer

Reviewer: make it so that this does change the working directory for the current shell. Hint: use `()`

Author: thanks for the heads up about this, i removed the section this was used in, but this allows documentation to be cleaner

3.  Upgrade to this commit or a later one.

4.  The environment variable 'AZUL_INDEXER_GOOGLE_SERVICE_ACCOUNT' has been

>     changed within `$project_root/environment`, run ::

Reviewer: Again, too much information.

>         cd $project_root

Reviewer: It's implied that this happens in the project root.

have too many entries in this file.

#2246 Add deployment incarnation counter
========================================

See upgrading instructions for #2143 below.

#2143 Merge service accounts for indexer and service
============================================================

1.  Before upgrading to this commit, run ::

        make -C terraform validate
        (cd terraform && terraform destroy -target=google_service_account.indexer)

2.  Upgrade to this commit or a later one.

3.  Let someone who can administer the SAM group that controls access to TDR

>     know of the rename for the service account email via Slack. The

Reviewer: ```suggestion
    know of the renamed service account via Slack. The
```

    administrator will need to replace the old service account email with the

>     new one. For example, ::

Reviewer: I don't think we should use a literal block for this. It looks odd:

![image](https://user-images.githubusercontent.com/5143256/95926671-b25f7d00-0d71-11eb-9580-c1aeed1c0088.png)

Author: I think following the same pattern that was used in [Section 1645](https://github.com/DataBiosphere/azul/blob/issues/amar/2143-merge-service-accounts/UPGRADING.rst#1645-rethink-template-config-variable-mechanism)  would work here. Need to see how github renders the changes.

Author: ![Screen Shot 2020-10-15 at 9 26 41 AM](https://user-images.githubusercontent.com/1947534/96158610-8acef880-0ec8-11eb-800f-c0cde2a208f0.png)

have too many entries in this file.

#2246 Add deployment incarnation counter
========================================

See upgrading instructions for #2143 below.

#2143 Merge service accounts for indexer and service
============================================================

1.  Before upgrading to this commit, run ::

        make -C terraform validate
        (cd terraform && terraform destroy -target=google_service_account.indexer)

2.  Upgrade to this commit or a later one.

3.  Let someone who can administer the SAM group that controls access to TDR
    know of the renamed service account via Slack. The administrator will need
    to replace the old service account email with the new one. For example,

>     Remove ::

Reviewer: ```suggestion
    replace ::
```

        azul-ucsc-indexer-foo@human-cell-atlas-travis-test.iam.gserviceaccount.com

>     Add ::

Reviewer: ```suggestion
    with ::
```

have too many entries in this file.

#2246 Add deployment incarnation counter
========================================

See upgrading instructions for #2143 below.

#2143 Merge service accounts for indexer and service
====================================================

1.  Before upgrading to this commit, run ::

>         make -C terraform validate

Reviewer: This is missing stuff. And the `-C terraform` followed by `(cd terraform` is a bit weird since both temporarily cd into that directory.

I don't know if this is correct but use it for inspiration:

```diff
Index: UPGRADING.rst
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- UPGRADING.rst	(revision 90c1b03c15994e1601105df9252be3cd86b05bb0)
+++ UPGRADING.rst	(date 1603338157551)
@@ -22,8 +22,12 @@
 
 1.  Before upgrading to this commit, run ::
 
-        make -C terraform validate
-        (cd terraform && terraform destroy -target=google_service_account.indexer)
+        source environment
+        _select foo
+        _preauth
+        cd terraform
+        make validate
+        terraform destroy -target=google_service_account.indexer
+        cd ..
 
 2.  Upgrade to this commit or a later one.
 
```

        (cd terraform && terraform destroy -target=google_service_account.indexer)

2.  Upgrade to this commit or a later one.

3.  Let someone who can administer the SAM group that controls access to TDR
    know of the renamed service account via Slack. The administrator will need
    to replace the old service account email with the new one. For example,

    replace ::

        azul-ucsc-indexer-foo@human-cell-atlas-travis-test.iam.gserviceaccount.com

    with ::

>         azul-ucsc-0-foo@human-cell-atlas-travis-test.iam.gserviceaccount.com

Reviewer: This can't work. At this point this account doesn't exist yet so the SAM UI will refuse to add it to the group.

Author: I've been able to add accounts to JadeStewards-dev that are not registered with SAM. The UI does not block me in JadeStewards-dev, although it does pop up this notification. I've not run into issues with authorization/authentication when adding the account into the JadeStewards-dev group before performing a deployment that changes service accounts.

Although if someone changes this at the Broad our documentation could be misleading or worse broken. I think moving this as a second to last step to complete would be ok, then informing the user to confirm access with `make -C terraform sam`

![Screen Shot 2020-10-22 at 9 26 17 AM](https://user-images.githubusercontent.com/1947534/96901903-b2820b80-1448-11eb-8aa0-676c23c07b81.png)

Author: 092b742 resolves this.

have too many entries in this file.

> #2246 Add deployment incarnation counter

Author: Keep this heading, in the body refer to the next instruction section

> ============================================================

Reviewer: too long

Reviewer: https://github.com/DataBiosphere/azul/pull/2210/files#r497967954

> This PR adds a new environemnt variable `$AZUL_DEPLOYMENT_INCARNATION`. Service

Reviewer: Too much information. Focus on the actual steps. No need to explain why. That new variable does not need to mentioned here.

account ids interpolate this value to ensure registration with TDR.

1.  Let someone who can administer authorization of SAM/TDR know of the rename
    for the service account via Slack. The administrator will need to replace
    the old service account_id with the new one. The change in the account_id
    adds the value of the `$AZUL_DEPLOYMENT_INCARNATION` to the account id.

        from:
            azul-ucsc-dev@human-cell-atlas-travis-test.iam.gserviceaccount.com
        to:
            azul-ucsc-0-dev@human-cell-atlas-travis-test.iam.gserviceaccount.com

    Note performing these steps
    with out the manual authorization from the administrator can lead to
    deployment errors during ``make deploy``.

2.  Environment variables will need to be re-sourced to pickup changes, run ::

        source environment
        _select $deployment
        _preauth

3.  Redeploy the upgraded infrastructure, run ::

        make package
        make deploy

> #2143 Merge service accounts for indexer and service

Author: combine with above section
